File: controllers/api/vendo_status_controller.py
from typing import Union

from fastapi import Depends
from fastapi.encoders import jsonable_encoder
from fastapi.responses import JSONResponse

from repository.vendo_status_repository import VendoStatusRepository


class VendoStatusController:
    def __init__(
            self,
            repository: VendoStatusRepository = Depends(VendoStatusRepository)
    ):
        self._repository = repository

    def search(self, vendo_id: Union[int, None] = None, from_date: Union[str, None] = None,
               to_date: Union[str, None] = None,
               active_only: bool = False) -> Union[dict, JSONResponse]:
        result = self._repository.get_hourly_status(vendo_id=vendo_id,
                                                    from_date=from_date,
                                                    to_date=to_date,
                                                    active_only=active_only)

        return JSONResponse({
            "data": jsonable_encoder(result)
        })

File: server/internal/controllers/vendo_status_controller.go
package controllers

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/jariesdev/vendoreport/internal/repository"
)

// VendoStatusController handles vendo status history endpoints.
type VendoStatusController struct {
	statusRepo repository.VendoStatusRepositoryInterface
	userRepo   repository.UserRepositoryInterface
}

func NewVendoStatusController(statusRepo repository.VendoStatusRepositoryInterface, userRepo repository.UserRepositoryInterface) *VendoStatusController {
	return &VendoStatusController{statusRepo: statusRepo, userRepo: userRepo}
}

// Search handles GET /vendo-status-history.
// Query params: vendo_id, from_date (YYYY-MM-DD), to_date, active_only (bool).
func (v *VendoStatusController) Search(c *gin.Context) {
	vendoIDStr := c.Query("vendo_id")
	fromDate := c.Query("from_date")
	toDate := c.Query("to_date")
	activeOnly, _ := strconv.ParseBool(c.Query("active_only"))

	var vendoIDPtr *uint
	if vendoIDStr != "" {
		if id, err := strconv.ParseUint(vendoIDStr, 10, 64); err == nil {
			uid := uint(id)
			vendoIDPtr = &uid
		}
	}
	var fromPtr, toPtr *string
	if fromDate != "" {
		fromPtr = &fromDate
	}
	if toDate != "" {
		toPtr = &toDate
	}

	ids, err := assignedVendoIDs(c, v.userRepo)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"detail": err.Error()})
		return
	}

	rows, err := v.statusRepo.GetHourlyStatus(vendoIDPtr, fromPtr, toPtr, activeOnly, ids)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"detail": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"data": rows})
}
